### ISSUE TITLE
v1.33.0 disabled extension appears to be running

### ISSUE THREAD

**Issue by john-aws** [Microsoft/vscode#71752]
Issue Type: <b>Bug</b>

I can reliably reproduce this problem. It started happening today, following today's VS Code update (March 2019, 1.33.0). It is closely related to #21930.

I have `formatOnSave=false` but I am suddenly seeing my React JS files being modified upon save (the JSX is being mangled). I have the [vscode-js-css-html-formatter](https://github.com/lonefy/vscode-js-css-html-formatter) extension installed, but disabled. Somehow, it appears to be functional (and also abusing `formatOnSave=false`).

The symptoms seem to suggest that the disabled extension is not actually disabled. If I uninstall the extension there is no indication that I need to reload VS Code to complete the uninstallation yet the save-on-format behavior is still present, post-uninstall.

If I uninstall the extension and actually reload VS Code, then the format-on-save behavior has gone and everything is good.

Any chance that v1.33.0 introduced a bug allowing disabled extensions, and uninstalled extension but VS-Code-not-yet-reloaded, to continue to run? Or is this extension somehow escaping whatever sandbox you have for disabled extensions?

This problem does **not** repro on v1.32.3.

VS Code version: Code 1.33.0 (0dd516dd412d42323fc3464531b1c715d51c4c1a, 2019-04-04T15:11:32.076Z)
OS version: Darwin x64 16.7.0

**Comment by john-aws** [comment]
Just noticed this bug raised against the extension: [Formatter must unregister when being disabled](https://github.com/Lonefy/vscode-JS-CSS-HTML-formatter/issues/91)

**Comment by sandy081** [comment]
@john-aws I see reload action when I uninstall/disable the extension (vscode-JS-CSS-HTML-formatter). Can you please check if the extension is shown when you run the command `Show Running Extensions` before you disbaled/uninstalled it?

**Comment by jrieken** [comment]
> The symptoms seem to suggest that the disabled extension is not actually disabled.

How did you disable it? In the sense of disabling the extension or disabling it with its setting?

**Comment by szhu** [comment]
Ran across this issue as well. I have a bug report with system info at #71769.

@sandy081 The disabled-but-not-really-disabled extension doesn't show up under `Show Running Extensions`.

@jrieken I am alleviating this issue temporarily by just uninstalling the extensions I don't need. I guess if you have some extensions activated for only certain workspaces, you could be in a tough spot.

**Comment by john-aws** [comment]
@sandy081 confirmed that the extension does *not* show up under `Show Running Extensions` after disabling but before uninstalling. In this state it does still appear to be reformatting my JSX.

@jrieken to disable the extension, I used the native VS Code Extension Manager, clicked the settings cog, and chose Disable (not Disable (Workspace)).

**Comment by sandy081** [comment]
If it is not shown under running extensions, the extension should not be running unless there is a bug in it.

@alexandrudima any insights on this?

**Comment by john-aws** [comment]
Is it's technically possible for a disabled extension to run (I don't know for sure if it is, but it seems to be), shouldn't VS Code exert more control? A clear preference from the user to disable an extension would ideally take the decision out of the hands of the extension writer.

**Comment by sandy081** [comment]
@john-aws It should not happen. When a user disables an extension and reload the window, it should not run. Can you please provide the steps for me to reproduce, also the set up (extensions you have) would be helpful too?

**Comment by szhu** [comment]
@sandy081 my #71752 has a specific repro steps.
(sorry if you were already aware -- just making sure!)

**Comment by sergei-dyshel** [comment]
There is a bug indeed.

I'm using this feature heavily (disabling of extensions globally/per workspace). For example [clangd](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd) extension appears to be still running (partially at least) while being disabled globally. There are several indications for that:
1. At least some LSP providers are acting (definition/formatting/references).
2. *Process Explorer* shows LSP executable under *Extension host*.
3. There is "Clang language server" output panel.

However the extension is not listed in *Show running extensions*. Also I found the following lines in *Log (Extension Host)* output panel:
```
[2019-04-15 13:48:56.421] [exthost] [info] ExtensionService#_doActivateExtension llvm-vs-code-extensions.vscode-clangd {"startup":false,"activationEvent":"onLanguage:c"}
[2019-04-15 13:48:56.421] [exthost] [info] ExtensionService#loadCommonJSModule /home/sergei/.vscode/extensions/llvm-vs-code-extensions.vscode-clangd-0.0.12/out/src/extension
```

Don't know what they mean though.

**Comment by sandy081** [comment]
@alexandrudima 

Today @jrieken has seen similar issue and we debugged it together. Outcome is that we are starting the extension host with all scanned extensions and not filtering the disabled extensions. See following code

https://github.com/Microsoft/vscode/blob/f7b33b613ed05e3649da8fb93cec88dac7d85856/src/vs/workbench/services/extensions/electron-browser/extensionService.ts#L433-L440

Not sure why are we not filtering out disabled extensions if it is initial start?

**Comment by sandy081** [comment]
Reverted the fix. I digged deep and above fix is not correct as the bug is not on renderer side. Root cause is in extension host.

- We create extension host with all extensions
- We start it after with only enabled extensions
- We are not holding those activation events those come between ext host is created and ext host is started. This is causing disabled extensions to activate.

Will come up with the accurate fix.
